logstash.go

```diff
@@ -30,8 +41,86 @@ func Init(logLevel string, logFileName string, env string, service string) bool
 	})
 
 	log.SetLevel(LogLevels[logLevel])
-	mw := io.MultiWriter(os.Stdout, logFile)
-
-	log.SetOutput(mw)
+	setLogFile(logFile)
+	go rotate(logFileName)
 	return true
 }
+
+func setLogFile(writer io.Writer) {
+	log.SetOutput(io.MultiWriter(os.Stdout, writer))
+}
+
+// rotate checks periodically if the
+// when the current logfile
+func rotate(logFileName string) {
+	log.Debug("starting rotation for", logFileName)
+
+	if stopRotation != nil {
+		log.Debug("stopping rotation")
+		close(stopRotation)
+	}
+	stopRotation = make(chan struct{})
+
+	ticker := time.NewTicker(rotationCheckInterval)
+	for {
+		select {
+		case <-stopRotation:
+			log.Debug("stopping rotation for", logFileName)
+			ticker.Stop()
+			return
+		case <-ticker.C:
+			f, err := os.Stat(logFileName)
+			if f.Size() > maxLogFileSize {
+				log.Debugf("log file too large, rotating the log file")
+				rotatedFilename := logFileName + ".1"
+
+				// if the rotated file already exist, add timestamp to its name and archive it
+				if _, err = os.Stat(rotatedFilename); err == nil {
+
+					err = delayedCompression(rotatedFilename)
+					if err != nil {
+						// if the renaming for compression failed, try again before rotating
+						continue
+					}
+				}
+
+				err = os.Rename(logFileName, rotatedFilename)
+				if err != nil {
+					log.Error("couldn't rename log file", logFileName, err)
+					continue
+				}
+
+				NewLogFile, err := os.Create(logFileName)
+				if err != nil {
+					panic(err)
+				}
+
+				setLogFile(NewLogFile)
```

Comment:
I think the setLogFile should be run asap at the top of the rotation, before any compression can take place, so the writes go to new file with minimal overlap.

rename, recreate -> do whatever

Reply:
This order is deliberate for error handling reasons.
If for one reason or another moving a file failed, The best way to ensure we're still writing logs and not losing/overwriting any, is to do nothing.
The only downside is that we could have logs for a few more milliseconds in the logfile to be rotated. I don't think it would make any difference

Reply:
ok, it makes sense now.

i got it wrong a little,

the path is just slightly backwards :)

do whatever -> rename -> recreate
👍

Reply:
let's lower the filesize limit to few k (20-30kb) and put this on dev ot test

Reply:
done